### PR TITLE
Support TIMESTAMP type in BigQueryIO with BEAM_ROW output type, and in read_gbq

### PR DESCRIPTION
**Pull request by svetakvsundhar**
Issue: https://github.com/apache/beam/issues/22909

This PR adds support for the `TIMESTAMP` time from BigQuery, and maps it to a `BEAM_ROW` type. This will be supported in Python's BigQuery IO implementation, as well as read_gbq.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by tvalentyn** [comment]
Run Python 3.8 PostCommit

**Comment by tvalentyn** [comment]
will defer to @ahmedabu98 for the review assuming tests pass.

File: sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py [hunk 1]
```diff
@@ -258,6 +258,18 @@ def test_table_schema_retrieve_with_direct_read(self):
               utype(id=4, name='customer2', type='test')
           ]))
 
+  @pytest.mark.it_postcommit
+  def test_table_schema_retrieve_with_timestamp(self):
+    with beam.Pipeline(argv=self.args) as p:
+      result = (
+          p | apache_beam.io.gcp.bigquery.ReadFromBigQuery(
+              gcs_location="gs://bqio_schema_test",
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py:266]: which project does this bucket live in?

Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py:266]: nvm, looks like apache-beam-testing

**Comment by svetakvsundhar** [comment]
whoops, i didnt realize the reviewer would be auto-assigned.

anyways, this integration test will fail with ` AttributeError: 'datetime.datetime' object has no attribute 'micros' [while running 'ReadFromBigQuery/ParDo(BeamSchemaConversionDoFn)']`

I wanted to ask, should `apache_beam.utils.timestamp` be using `TimestampCoder`? Looks like it's using `RowCoder` now.

**Comment by svetakvsundhar** [comment]
cc: @TheNeuralBit

File: sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py [hunk 2]
```diff
@@ -258,6 +258,18 @@ def test_table_schema_retrieve_with_direct_read(self):
...
+              dataset="beam_bigquery_io_test",
+              table="taxi",
+              project="apache-beam-testing",
+              output_type='BEAM_ROW'))
+      assert_that(result, equal_to([]))
```
Comment (ahmedabu98) [inline sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py:271]: What is this testing for?

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Thanks @svetakvsundhar, left a couple comments

**Comment by tvalentyn** [comment]
waiting on author

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PreCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PreCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py [hunk 3]
```diff
@@ -258,6 +258,123 @@ def test_table_schema_retrieve_with_direct_read(self):
...
+                      tzinfo=datetime.timezone.utc),
+                  ride_id='0004b5de-8db1-425b-8eec-7a25b04a1ee0',
+                  point_idx=9,
+                  latitude=40.757380000000005,
+                  timestamp='2021-11-05T16:57:43.60906-04:00',
+                  meter_reading=0.37846154,
+                  meter_increment=0.04205128,
+                  ride_status='enroute',
+                  passenger_count=1,
+                  longitude=-73.98969000000001),
```
Comment (ahmedabu98) [inline sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py:299]: Is it possible to only test for the timestamp field since that's what this test is focusing on? It would help make the test more concise and easy to read in the future. 

One way is to change it so that ReadFromBigQuery queries for the timestamp field. If schema conversion isn't implemented yet for query reads, another way may be to add a step that filters `result` so you have a PCollection of timestamp elements.

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Thanks @svetakvsundhar, left a few comments

@tvalentyn does the use of `datetime.datetime` look good to you?

**Comment by svetakvsundhar** [comment]
> Thanks @svetakvsundhar, left a few comments
> 
> @tvalentyn does the use of `datetime.datetime` look good to you?

Actually, I think we'd need to further convert this into a an `apache_beam.utils.timestamp.Timestamp` object. fix incoming...

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py [hunk 4]
```diff
@@ -61,6 +61,25 @@ def test_check_conversion_with_empty_schema(self):
...
+    schema = bigquery.TableSchema(fields=fields)
+
+    usertype = bigquery_schema_tools.generate_user_type_from_bq_schema(
+        the_table_schema=schema)
+    self.assertEqual(
+        usertype.__annotations__,
+        {
+            'stn': typing.Optional[str],
+            'temp': typing.Sequence[np.float64],
+            'times': typing.Any
```
[the remark this answers is not in the document]
Comment (svetakvsundhar) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py:80]: I think because this is using the `.from_utc_datetime` method, and datetime isn't defined here https://github.com/apache/beam/blob/master/sdks/python/apache_beam/typehints/schemas.py#L23, the type is shunted to `typing.Any`

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py (hunk 3 above)
Comment (svetakvsundhar) [reply sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py]: thanks! edited to reflect in previous tests

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py:80]: Ahh I see it changed in [this commit](https://github.com/apache/beam/pull/23070/commits/6881243eec5b307dbb73b0905114547f463b1506#diff-a2c4c0db5641c542c2ce62959c0cca307d810109c982fb1065e123ec8426f296R45). [`from_utc_datetime()`](https://github.com/apache/beam/blob/8a1aad6824319abf662570de36e2580db8e9450e/sdks/python/apache_beam/utils/timestamp.py#L98) returns a `Timestamp` object. Would testing for Optional[Timestamp] work? As it is now the test for timestamp is not very accurate.

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py [hunk 5]
```diff
@@ -76,11 +79,11 @@ def generate_user_type_from_bq_schema(the_table_schema):
 
 
 def bq_field_to_type(field, mode):
-  if mode == 'NULLABLE':
+  if mode == 'NULLABLE' or mode is None or mode == '':
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:82]: what's the motivation behind this change? I wonder if it could be a breaking change for some user pipelines.

Comment (svetakvsundhar) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:82]: I wonder if it could be a breaking change for some user pipelines, because some hints downstream are more restrictive.

>>It won't be a breaking change since this solely applies to pipelines utilizing the `BEAM_ROW` output type (which none do yet). The motivation is to make the default mode Nullable ([as done in the Java sdk](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/model/TableFieldSchema.html#setMode-java.lang.String-)).

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (svetakvsundhar) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py:80]: Hm so when I look at `dict_of_tuples`, I see `[('stn', typing.Optional[str]), ('temp', typing.Sequence[numpy.float64]), ('times', typing.Optional[<bound method Timestamp.from_utc_datetime of <class 'apache_beam.utils.timestamp.Timestamp'>>])]` which essentially becomes `Typing.Any` when calling `named_fields_to_schema` (https://github.com/apache/beam/blob/master/sdks/python/apache_beam/typehints/schemas.py#L120) 


IIUC this is intended

[the remark this answers is not in the document]
Comment (svetakvsundhar) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py:80]: @tvalentyn , @TheNeuralBit for any thoughts

[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py:80]: ` typing.Optional[<bound method Timestamp.from_utc_datetime of <class 'apache_beam.utils.timestamp.Timestamp'>>] `
This doesn't sound correct. Why are we not using  apache_beam.utils.timestamp.Timestamp ?

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py [hunk 6]
```diff
@@ -38,8 +38,10 @@
     "STRING": str,
     "INTEGER": np.int64,
     "FLOAT64": np.float64,
+    "FLOAT": np.float64,
     "BOOLEAN": bool,
     "BYTES": bytes,
+    "TIMESTAMP": datetime.datetime
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py]: Should the type here be  `apache_beam.utils.timestamp.Timestamp` ?

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py [hunk 7]
```diff
@@ -38,8 +39,10 @@
     "STRING": str,
     "INTEGER": np.int64,
     "FLOAT64": np.float64,
+    "FLOAT": np.float64,
     "BOOLEAN": bool,
     "BYTES": bytes,
+    "TIMESTAMP": apache_beam.utils.timestamp.Timestamp.from_utc_datetime
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:45]: Should the type here be `apache_beam.utils.timestamp.Timestamp` ?

Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:45]: cc: @Abacn @TheNeuralBit who worked on  https://github.com/apache/beam/blob/master/sdks/python/apache_beam/typehints/schemas.py#L36-L37 

Folks, do you know what (if any) aspects we need to take care of for this change to correctly work with schemas & xlang?

Comment (Abacn) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:45]: I had some discussion with @svetakvsundhar before on the question whether implement TIMESTAMP as Python's native `datetime.datetime` or beam's `apache_beam.utils.timestamp.Timestamp`. Both should work within the Python SDK, but `datetime.datetime` is not portable. So I suggested the latter. In a xlang scenario it should just work.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:45]: Thanks. It seems that we should drop `.from_utc_datetime` as that is a method, which returns the Timestamp.
Any concerns with that @svetakvsundhar ?

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheNeuralBit) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools_test.py:80]: It looks like somewhere we're creating a type `Optional[Timestamp.from_utc_datetime]`, which is not a real typehint, as Valentyn indicated it should be `Optional[Timestamp]`

File: sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py (hunk 7 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/io/gcp/bigquery_schema_tools.py:45]: +1

```suggestion
    "TIMESTAMP": apache_beam.utils.timestamp.Timestamp
```

should do it

**Comment by svetakvsundhar** [comment]
Thanks all for the suggestions. Because the ` BQ Timestamp` object was actually a `datetime.datetime`, we needed to modify the conversion ParDo logic to convert into an `apache_beam_utils.timestamp.Timestamp`

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Comment by svetakvsundhar** [comment]
Run Python PreCommit

**Comment by svetakvsundhar** [comment]
Run Python 3.7 PostCommit

**Review by tvalentyn — APPROVED** [review]

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

**Comment by tvalentyn** [comment]
@AnandInguva is looking into py3.10 issue, looks unrelated, py3.8 has bq tests and should provide sufficient coverage, not sure why py3.7 failed, looks like infra error

**Comment by tvalentyn** [comment]
other errors:
apache_beam.runners.portability.fn_api_runner.fn_runner_test.FnApiRunnerTestWithGrpcAndMultiWorkers.test_batch_pardo_trigger_flush  --> https://github.com/apache/beam/issues/21104
apache_beam.runners.portability.portable_runner_test.PortableRunnerTestWithExternalEnv.test_pardo_large_input -> https://github.com/apache/beam/issues/22321

**Comment by ahmedabu98** [comment]
Py37 postcommits have been consistently timing out: `Build timed out (after 240 minutes). Marking the build as aborted.`

**Comment by ahmedabu98** [comment]
Py38 more reliable as it isn't responsible for as many tests as Py37

**Review by ahmedabu98 — APPROVED** [review]
This LGTM, let's wait for Py38 Postcommits though

**Comment by svetakvsundhar** [comment]
sg, looks like 3.7 ran the tests in question 
Screen Shot 2022-10-19 at 1 13 11 PM

**Comment by svetakvsundhar** [comment]
sg, looks like 3.7 ran the tests in question 
Screen Shot 2022-10-19 at 2 41 37 PM

**Comment by tvalentyn** [comment]
> Py37 postcommits have been consistently timing out: Build timed out (after 240 minutes). Marking the build as aborted.

This is strange, I have increased the timeout in: https://github.com/apache/beam/pull/23595
